### PR TITLE
OGM-1238 Keep order of collections of elements in MongoDB

### PR DESCRIPTION
**Pull request by DavideD**
https://hibernate.atlassian.net/browse/OGM-1238

The mongodb driver returns collection of elements in the same order they appear in the DB, we should do the same.

### CODE REVIEW

**Comment by DavideD** [comment]
Jenkins, retest this please

File: core/src/main/java/org/hibernate/ogm/datastore/document/association/spi/AssociationRows.java [hunk 1]
```diff
@@ -28,7 +27,7 @@
 	private final Map<RowKey, AssociationRow<?>> rows;
 
 	public AssociationRows(AssociationKey associationKey, Collection<?> wrapped, AssociationRowFactory associationRowFactory) {
-		this.rows = newHashMap( wrapped.size() );
+		this.rows = new LinkedHashMap<>( wrapped.size() );
```
Comment (gsmet) [inline core/src/main/java/org/hibernate/ogm/datastore/document/association/spi/AssociationRows.java]: Please add a new `newLinkedHashMap` method to our CollectionHelper.

I'll work on pushing the changes we made in HV regarding the sizing.

(With the current code, you always undersize the map)

File: core/src/main/java/org/hibernate/ogm/model/spi/Association.java [hunk 2]
```diff
@@ -184,7 +184,7 @@ public int size() {
 		}
 		else {
 			// It may be a bit too large in case of removals, but that's fine for now
-			Set<RowKey> keys = new HashSet<RowKey>( cleared ? currentState.size() : snapshot.size() + currentState.size() );
+			Set<RowKey> keys = new LinkedHashSet<RowKey>( cleared ? currentState.size() : snapshot.size() + currentState.size() );
```
Comment (gsmet) [inline core/src/main/java/org/hibernate/ogm/model/spi/Association.java]: Same here. Let's create a `newLinkedHashSet` method in `CollectionHelper` and I'll optimize it.

File: mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/embeddable/ElementCollectionOrderWithSessionTest.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+import org.hibernate.ogm.utils.TestForIssue;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+/**
+ * @author Davide D'Alto
+ */
+@TestForIssue(jiraKey = "OGM-1238")
+public class ElementCollectionOrderWithSessionTest extends OgmTestCase {
```
Comment (gsmet) [inline mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/embeddable/ElementCollectionOrderWithSessionTest.java]: Is it really worth it to add 2 tests for that?

Comment (DavideD) [inline mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/embeddable/ElementCollectionOrderWithSessionTest.java]: I think so, we are testing two different interfaces and the test classes are simple. I don't see where's the problem. Potentially, comsething could happen between EntityManager and Session that causes a test to fail.

File: core/src/main/java/org/hibernate/ogm/datastore/document/association/spi/AssociationRows.java (hunk 1 above)
Comment (DavideD) [inline core/src/main/java/org/hibernate/ogm/datastore/document/association/spi/AssociationRows.java]: Perhaps, but this is preexisting code, why stopping this PR that tries to solve a different issue?

**Comment by DavideD** [comment]
Closing this PR until I fixed a couple of things

File: mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/embeddable/ElementCollectionOrderWithSessionTest.java (hunk 3 above)
Comment (gsmet) [inline mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/embeddable/ElementCollectionOrderWithSessionTest.java]: OK, it's just that I wasn't sure it was useful.

**Comment by DavideD** [comment]
jenkins, retest this please

**Pull request closed without merging**